### PR TITLE
[SDP-894] data/serve: Two wallet providers in the same address

### PR DESCRIPTION
**Pull request by ceciliaromao**
### What

Supports the receiver to register a second (or more) wallet provider in the same address to receive another disbursement.

### Why

If a receiver has two payments using two different wallet providers and they have the same address, only one can get registered

### Known limitations

N/A

#### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).
* [x] This PR's title starts with the name of the package, area, or subject affected by the change.

#### Thoroughness

* [x] This PR adds tests for the new functionality or fixes.
* [x] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [x] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [x] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [x] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [x] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [x] This is not a breaking change.
* [x] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [ ] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

**Review by CaioTeixeira95 — APPROVED** [review]
Thank you for working on this! 🚀 

My only last comment is to add a new test for the `VerifyReceiverRegistrationHandler.VerifyReceiverRegistration` handler to ensure that we can add different wallet providers with the same wallet address.

File: internal/serve/httphandler/receiver_registration.go [hunk 1]
```diff
@@ -52,7 +52,7 @@ func (h ReceiverRegistrationHandler) ServeHTTP(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	rw, err := h.ReceiverWalletModel.GetByStellarAccountAndMemo(ctx, sep24Claims.SEP10StellarAccount(), sep24Claims.SEP10StellarMemo())
+	rw, err := h.ReceiverWalletModel.GetByStellarAccountAndMemo(ctx, sep24Claims.SEP10StellarAccount(), sep24Claims.SEP10StellarMemo(), sep24Claims.ClientDomain())
```
Comment (marwen-abid) [inline internal/serve/httphandler/receiver_registration.go:55]: I was looking into `sep24Claims.Valid()` to see if `ClientDomain()` is required there and it seems like it's not. 

The current test only checks that it's valid when not empty. 
```go
if c.ClientDomain() != "" {
  err = utils.ValidateDNS(c.ClientDomain())
  if err != nil {
	  return fmt.Errorf("client_domain is invalid: %w", err)
  }
}
```
	
This means that it's possible for ClientDomain to be empty. 

Can you please check when ClientDomain is empty?
	
And if it's acceptable that client domain is empty, maybe we should make it an optional  querying field in `GetByStellarAccountAndMemo`.

File: internal/data/receivers_wallet.go [hunk 2]
```diff
@@ -461,12 +461,13 @@ func (rw *ReceiverWalletModel) GetByStellarAccountAndMemo(ctx context.Context, s
 		FROM receiver_wallets rw
 		JOIN wallets w ON rw.wallet_id = w.id
 		WHERE rw.stellar_address = $1
+		AND w.sep_10_client_domain = $2
 	`
 
 	// append memo to query if it is not empty
```
Comment (marwen-abid) [inline internal/data/receivers_wallet.go:467]: Related to my other comment. If ClientDomain can be empty, we should probably switch this to something like this: 

```go
         args := []interface{}{stellarAccount}
	if stellarMemo != "" {
		query += " AND rw.stellar_memo = $2"
		args = append(args, stellarMemo)
	} else {
		query += " AND (rw.stellar_memo IS NULL OR rw.stellar_memo = '')"
	}
	
	if clientDomain != "" {
		query += " AND w.sep_10_client_domain = $3"
		args = append(args, clientDomain)
	}
	```

File: internal/data/receivers_wallet.go [hunk 3]
```diff
@@ -441,7 +441,7 @@ func (rw *ReceiverWalletModel) UpdateStatusByDisbursementID(ctx context.Context,
 }
 
 // GetByStellarAccountAndMemo returns a receiver wallets that match the Stellar Account.
-func (rw *ReceiverWalletModel) GetByStellarAccountAndMemo(ctx context.Context, stellarAccount, stellarMemo string) (*ReceiverWallet, error) {
+func (rw *ReceiverWalletModel) GetByStellarAccountAndMemo(ctx context.Context, stellarAccount, stellarMemo string, clientDomain string) (*ReceiverWallet, error) {
```
Comment (marwen-abid) [inline internal/data/receivers_wallet.go:444]: nit: 
```suggestion
func (rw *ReceiverWalletModel) GetByStellarAccountAndMemo(ctx context.Context, stellarAccount, stellarMemo, clientDomain string) (*ReceiverWallet, error) {
```

**Review by marwen-abid — CHANGES_REQUESTED** [review]
My only comments are related to `ClientDomain` possibly being empty. Please check if that's case and if it is, we need to handle that correctly.

**Review by marwen-abid — APPROVED** [review]
LGTM. Thanks for addressing the comments

**Pull request merged**